rule/identical-branches.go

```diff
@@ -0,0 +1,82 @@
+package rule
+
+import (
+	"go/ast"
+
+	"github.com/mgechev/revive/lint"
+)
+
+// IdenticalBranchesRule warns on constant logical expressions.
+type IdenticalBranchesRule struct{}
+
+// Apply applies the rule to given file.
+func (r *IdenticalBranchesRule) Apply(file *lint.File, _ lint.Arguments) []lint.Failure {
+	var failures []lint.Failure
+
+	onFailure := func(failure lint.Failure) {
+		failures = append(failures, failure)
+	}
+
+	astFile := file.AST
+	w := &lintIdenticalBranches{astFile, onFailure}
+	ast.Walk(w, astFile)
+	return failures
+}
+
+// Name returns the rule name.
+func (r *IdenticalBranchesRule) Name() string {
+	return "identical-branches"
+}
+
+type lintIdenticalBranches struct {
+	file      *ast.File
+	onFailure func(lint.Failure)
+}
+
+func (w *lintIdenticalBranches) Visit(node ast.Node) ast.Visitor {
+	n, ok := node.(*ast.IfStmt)
+	if !ok {
+		return w
+	}
+
+	if n.Else == nil {
+		return w
+	}
+	branches := []*ast.BlockStmt{n.Body}
+
+	elseBranch, ok := n.Else.(*ast.BlockStmt)
+	if !ok { // if-else-if construction
+		return w
+	}
+	branches = append(branches, elseBranch)
+
+	if w.identicalBranches(branches) {
+		w.newFailure(n, "both branches of the if are identical")
+	}
+
+	return w
+}
+
+func (w *lintIdenticalBranches) identicalBranches(branches []*ast.BlockStmt) bool {
+	if len(branches) < 2 {
+		return false
+	}
+
+	ref := gofmt(branches[0])
+	for i := 1; i < len(branches); i++ {
+		if gofmt(branches[i]) != ref {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (w lintIdenticalBranches) newFailure(node ast.Node, msg string) {
+	w.onFailure(lint.Failure{
```

Comment:
Do you think there's a case when we'll return a false positive? If we have a variable declaration with the same name, but different types, for example, are we going to catch this?

Since you're using `gofmt`, I don't think we can hit such a case, but just pointing this out.

Reply:
The rule uses (inefficient) string comparison to find identical branches so false positives are not possible.
In the near future I'll try to provide a, I hope, more efficient implementation based on AST comparison.